astar_utils/spectral_types.py

```diff
@@ -0,0 +1,150 @@
+    individual attribute, creation can only happen via the constructor string
+    described above.
+
+    One of the main features of this class (and indeed motivation for its
+    creation in the first place) is the ability to compare, and thus order,
+    instances of the class based on their two-component spectral (sub)class.
+    In this context, a "later" spectral type is considered "greater" than an
+    "earlier" one, i.e. O < B < A < F < G < K < M, which is consistent with the
+    convention of numerical subtypes, where 0 < 9 holds true. This is, however,
+    in contrast to the physical meaning of this parameter, which is correlated
+    with stellar effective temperature in reverse order, meaning T(A) > T(G)
+    and T(F0) > T(F9), by convention. On the other hand, in many visualisations
+    such as the Herzsprung-Russel diagram, it is common practice to represent
+    temperature on the x-axis in descending order, meaning a sorted list of
+    instances of this class will already have the typical order of such
+    diagrams.
+
+    In this context, the luminosity class (if any) is ignored for sorting and
+    comparison (<, >, <=, >=), as it represents a second physical dimension.
+    However, instances of this class may also be compared for equality (== and
+    !=), in which case all three attributes are considered.
+
+    Attributes
+    ----------
+    spectral_class : str
+        Main spectral class (OBAFGKM).
+    spectral_subclass : str or None
+        Numerical spectral subclass (0.0-9.9).
+    luminosity_class : str or None
+        Roman numeral luminosity class (I-V).
+
+    Examples
+    --------
+    >>> from astar_utils import SpectralType
+    >>> spt = SpectralType("A0V")
+    >>> spt.spectral_class
+    'A'
+
+    >>> spt.spectral_subclass
+    0.0
+
+    >>> spt.luminosity_class
+    'V'
+
+    >>> spts = [SpectralType(s) for s in
+    ...         ["G2", "M4.0", "B3", "F8", "K6.5",
+    ...          "A", "A0", "A9", "O8"]]
+    >>> sorted(spts)  # doctest: +NORMALIZE_WHITESPACE
+    [SpectralType('O8'),
+     SpectralType('B3'),
+     SpectralType('A0'),
+     SpectralType('A'),
+     SpectralType('A9'),
+     SpectralType('F8'),
+     SpectralType('G2'),
+     SpectralType('K6.5'),
+     SpectralType('M4')]
+
+    Note that a missing spectral subtype is considered as 5 (middle of the
+    main spectral class) in the context of sorting, as shown in the example
+    with the spectral type "A" ending up between "A0" and "A9".
+    """
+
+    spectral_class: str = field(init=False, default="")
+    spectral_subclass: float | None = field(init=False, default=None)
+    luminosity_class: str | None = field(init=False, default=None)
+    spectype: InitVar[str]
+    _cls_order: ClassVar = "OBAFGKM"  # descending Teff
+    _regex: ClassVar = re.compile(
+        r"^(?P<spec_cls>[OBAFGKM])(?P<sub_cls>\d(?:\.\d)?)?"
+        "(?P<lum_cls>I{1,3}|IV|V)?$", re.A | re.I)
```

Comment:
Pleases remove the flags.
```suggestion
        "(?P<lum_cls>I{1,3}|IV|V)?$")
```

The ASCII flag does not seem to add value, why not allow other unicode numerals? If we keep it, please spell it out, `re.ASCII`, as I had to look up what it means.

The IGNORECASE flag is broken:
```
>>> type1 = SpectralType("a")
>>> type2 = SpectralType("G")
>>> type1 < type2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "[...]/astar-utils/astar_utils/spectral_types.py", line 144, in __l
t__
    return self._comp_tuple < other._comp_tuple
  File "[...]/astar-utils/astar_utils/spectral_types.py", line 138, in _co
mp_tuple
    return (self._cls_order.index(self.spectral_class), sub_cls)
ValueError: substring not found
```

Reply:
OK, please keep the `re.ASCII` as this is not particularly useful I think:
```
assert SpectralType("G᱆.᱁") == SpectralType("G6.1")
```

Reply:
Yeah I think if non-ASCII is passed, there's something else wrong somewhere, so probably good to catch that...

Reply:
I originally put the IGNORECASE flag in because I thought it should be fine to pass e.g. "a0v", which I remember seeing a few times in some places (Templates I think), and I see no reason that shouldn't be allowed. BUT (and that was my original idea but I forgot to add this) then there should probably be an `.upper()` somewhere! I'd say right at the assignment of the internal attributes probably is the best place. Then any lowercase input would still parse, but be converted to uppercase immediately. And ofc also then add tests for those "cases" 😛

Reply:
Also fine with me to add the `.upper()` somewhere. 

I was trying to mess things up with `A2İ`, that is, with a Turkish captial i. But that does not work, because it is not ASCII, and because the luminosity class is never used for anything. Removing `re.A` does actually make `A2İ` match with `re.I` though.